File: microservice-architecture/service-communication/js/test/test_communication.js
import {
  AsyncOrderService,
  EventBus,
  InventoryService,
  OrderStore,
  PaymentService,
  registerAsyncWorkflow,
  SynchronousOrderService
} from '../src/communication.js'

function assertEqual(expected, actual, message) {
  if (expected !== actual) {
    throw new Error(`${message} expected=${expected} actual=${actual}`)
  }
}

const syncService = new SynchronousOrderService(
  new InventoryService({ 'SKU-BOOK': 5 }),
  new PaymentService()
)
assertEqual('CREATED', syncService.placeOrder('ORD-1001', 'SKU-BOOK', 2).status, 'sync order')

const bus = new EventBus()
const store = new OrderStore()
registerAsyncWorkflow(bus, store, new InventoryService({ 'SKU-BOOK': 5 }), new PaymentService())
const asyncService = new AsyncOrderService(bus, store)
assertEqual('PENDING', asyncService.placeOrder('ORD-2001', 'SKU-BOOK', 2).status, 'async pending')
bus.drain()
assertEqual('CREATED', store.get('ORD-2001').status, 'async created')

const failingBus = new EventBus()
const failingStore = new OrderStore()
registerAsyncWorkflow(
  failingBus,
  failingStore,
  new InventoryService({ 'SKU-BOOK': 5 }),
  new PaymentService(['ORD-2002'])
)
new AsyncOrderService(failingBus, failingStore).placeOrder('ORD-2002', 'SKU-BOOK', 1)
failingBus.drain()
assertEqual('PAYMENT_FAILED', failingStore.get('ORD-2002').status, 'async failure')

console.log('service-communication(js) tests passed')

File: microservice-architecture/service-communication/python/test/test_communication.py
import os
import sys
import unittest

sys.path.append(os.getcwd())

from src.communication import (  # noqa: E402
    AsyncOrderService,
    EventBus,
    InventoryService,
    OrderStore,
    PaymentService,
    SynchronousOrderService,
    register_async_workflow,
)


class ServiceCommunicationTest(unittest.TestCase):
    def test_synchronous_workflow(self) -> None:
        inventory = InventoryService({"SKU-BOOK": 5})
        payment = PaymentService()
        service = SynchronousOrderService(inventory, payment)

        order = service.place_order("ORD-1001", "SKU-BOOK", 2)
        self.assertEqual("CREATED", order.status)

    def test_asynchronous_workflow(self) -> None:
        inventory = InventoryService({"SKU-BOOK": 5})
        payment = PaymentService()
        bus = EventBus()
        store = OrderStore()
        register_async_workflow(bus, store, inventory, payment)

        service = AsyncOrderService(bus, store)
        initial = service.place_order("ORD-2001", "SKU-BOOK", 2)
        self.assertEqual("PENDING", initial.status)

        bus.drain()
        self.assertEqual("CREATED", store.get("ORD-2001").status)

    def test_asynchronous_payment_failure(self) -> None:
        inventory = InventoryService({"SKU-BOOK": 5})
        payment = PaymentService(["ORD-2002"])
        bus = EventBus()
        store = OrderStore()
        register_async_workflow(bus, store, inventory, payment)

        service = AsyncOrderService(bus, store)
        service.place_order("ORD-2002", "SKU-BOOK", 1)
        bus.drain()

        self.assertEqual("PAYMENT_FAILED", store.get("ORD-2002").status)


if __name__ == "__main__":
    unittest.main()
